xattr -> path based resource tracking and systemd-networkd

ddiss opened the pull request:
@scabrero : this continues on from your previous systemd-networkd adventures at https://github.com/rapido-linux/rapido/pull/147 . It's at the point where it should easily support the multi-bridge + multi-nic setup which you desired. There's still quite a bit to add + clean up, so I'm flagging it WIP for now.

TODOs can be found in the commit messages. I'm a few more tasks will pop up. High on my list is distro-kernel support, e.g. booting `/boot/vmlinuz` if `KERNEL_SRC` is undefined.

scabrero reviewed (commented):
Awesome! The ability to run more than 2 vms with multiple nics is a big step forward!

So far all these changes LGTM.

ddiss commented:
@scabrero new changes on top attempt to get rid of `IP_ADDR#`, `MAC_ADDR#` etc.
@luis-henrix you'll be sad to hear that I added back some funky awk ini parsing functionality, so that we can collect IP and MAC address configuration from the ini style systemd-networkd `.network` files.

Feels like it's getting a bit closer now. I think the main blocker aside from size / perf benchmarks is how to get this running alongside systemd - at the moment we always assume that systemd isn't installed. Similar priority is the helper script to migrate old rapido.conf `IP_ADDR#`, `MAC_ADDR#`, etc. settings to the net-conf directory.

ddiss commented:
I have some minimal changes to boot the locally running kernel from /boot if `KERNEL_SRC` isn't set. It works well with networked VMs now that we no longer rely on `ip=X` kernel cmdline.
I think it makes sense to keep the changes separate for now, so that this PR doesn't become too much of a dumping ground.

luis-henrix commented:
Ugh, I took a quick look and I didn't saw anything really shocking.  But this is a massive change and it'll need to wait some more time before I can really give it a try and do a more in-depth review :-(

ddiss commented:
I've benchmarked, squashed, reordered and cleaned things up to the point that I think the WIP flag can come down.

Single vNIC + static IP benchmarks show a ~1s hit to combine cut+boot runtime for the new `simple-network` test case, mostly spent in `systemd-networkd-wait-online`. Networked images grow by ~1M to accommodate for `systemd-networkd`.

Networkless images have a negligible change in size (no cut+boot runtime change), with the xattr -> path based resource tracking consuming a few hundred bytes.
@morbidrsa interested to get your thoughts too if you have some time :-)

Everything up to and including `cut/fstests: set memory based on FSTESTS_ZRAM_SIZE` could be split into a separate non-systemd-networkd PR, but given that the resource tracking changes will also break out-of-tree `cut` scripts I think it makes sense to bundle them all together.

ddiss commented:
minor updates:
- fix netd config paths in `_vm_ar_cfg_` helpers and lrbd autorun
- sign HEAD commit

ddiss commented on vm_autorun.env:
```diff
@@ -113,10 +151,117 @@ function _vm_ar_configfs_mount
+}
+
+_vm_ar_cfg_ips_nomask() {
+	_vm_ar_cfg_ips_nomask "$1" "$2" "nomask"
```
I'm still running with this patchset and just hit the endless loop here... Will squash and push a fix for this.

ddiss commented:
minor changes since previous version:
- fix ip address regex in lio_rbd
- fix `_vm_ar_cfg_ips_nomask` recursive call
- use local vars in net config helpers to avoid clobbering caller iterators

ddiss commented on cut/fstests_exfat.sh:
```diff
@@ -9,6 +9,10 @@ _rt_require_dracut_args "$RAPIDO_DIR/autorun/lib/fstests.sh" \
+_rt_human_size_in_b "${FSTESTS_ZRAM_SIZE:-1G}" zram_bytes \
+	|| _fail "failed to calculate memory resources"
+# 2x multiplier for one test and one scratch zram. +2G as buffer
+_rt_mem_resources_set "$((2048 + (zram_bytes * 2 / 1048576))M"
```
^ Ran into this today: missing trailing `)`. Will fix and squash in.

ddiss commented:
minor change since previous version:
- fix `cut/fstests_exfat.sh _rt_mem_resources_set` calculation

ddiss commented:
Any feedback here? I'd like to get this merged in the next week or so if there are no objections.

pevik commented:
@ddiss rebasing to current master could lower number of commits (4 commits behind) to make reviewing little bit easier. The rest looks to me related, but again, if anything unrelated can be merged first it'd help reviewing.

ddiss commented:
@morbidrsa @luis-henrix regarding what needs to be done for any out-of-tree cut/autorun scripts with this patchset, the new `simple-network` runner is provided as a before / after example:
```
> git diff f8873c2..4a98546 -- cut/simple_network.sh
diff --git a/cut/simple_network.sh b/cut/simple_network.sh
index 3304bb2..0331d36 100755
--- a/cut/simple_network.sh
+++ b/cut/simple_network.sh
@@ -6,12 +6,12 @@ RAPIDO_DIR="$(realpath -e ${0%/*})/.."
 . "${RAPIDO_DIR}/runtime.vars"
 
 _rt_require_dracut_args "$RAPIDO_DIR/autorun/simple_network.sh" "$@"
+_rt_require_networking
+_rt_cpu_resources_set "1"
+_rt_mem_resources_set "512M"
 
 "$DRACUT" \
-       --install "nc hostname ip ping" \
-       $DRACUT_RAPIDO_INCLUDES \
+       --install "nc hostname" \
        --modules "base" \
-       $DRACUT_EXTRA_ARGS \
+       "${DRACUT_RAPIDO_ARGS[@]}" \
        "$DRACUT_OUT" || _fail "dracut failed"
-
-_rt_xattr_vm_resources_set "$DRACUT_OUT" "1" "512M"
```

Explanation:
* cpu / memory / network resources are requested via the helpers *before* dracut invocation
  * these helpers add resource paths to the `DRACUT_RAPIDO_ARGS` array, instead of xattrs as is currently the case
* `ip` and `ping`  (along with `systemd-networkd`) are added as dracut install targets by `_rt_require_networking`, so needn't be explicitly listed
* `DRACUT_RAPIDO_INCLUDES` and `DRACUT_EXTRA_ARGS` have also been rolled into `DRACUT_RAPIDO_ARGS`

The only other thing to do is run `tools/net_conf_migrate.sh`, which will generate a VM configuration for `systemd-networkd` based on the current `MAC_ADDR`, `TAP_DEV`, `HOSTNAME`, `IP_ADDR[_DHCP]' rapido.conf settings.

ddiss commented:
I've pulled the trigger and merged this patchset, as I've been using it for quite some time and can't really imagine going back. Please feel free to raise tickets or ping me if you run into any issues, including with out-of-tree stuff.
